diff --git a/arch/x86/configs/gki_defconfig b/arch/x86/configs/gki_defconfig
index c6e83c40ca54..ac72293decd7 100644
--- a/arch/x86/configs/gki_defconfig
+++ b/arch/x86/configs/gki_defconfig
@@ -581,6 +581,8 @@ CONFIG_IIO_FORMAT_KUNIT_TEST=m
 CONFIG_POWERCAP=y
 CONFIG_IDLE_INJECT=y
 CONFIG_USB4=m
+CONFIG_USB4_DEBUGFS_WRITE=y
+CONFIG_USB4_DEBUGFS_MARGINING=y
 CONFIG_ANDROID_BINDER_IPC=y
 CONFIG_ANDROID_BINDERFS=y
 CONFIG_ANDROID_BINDER_IPC_RUST=m
diff --git a/arch/x86/configs/x86_64_defconfig b/arch/x86/configs/x86_64_defconfig
index 61e25f6209ed..9a9e0d4d95ad 100644
--- a/arch/x86/configs/x86_64_defconfig
+++ b/arch/x86/configs/x86_64_defconfig
@@ -267,6 +267,8 @@ CONFIG_SECURITY_SELINUX_BOOTPARAM=y
 CONFIG_SECURITY_SELINUX_DISABLE=y
 CONFIG_PRINTK_TIME=y
 CONFIG_DEBUG_KERNEL=y
+CONFIG_DEBUG_FS=y
+CONFIG_DEVMEM=y
 CONFIG_MAGIC_SYSRQ=y
 CONFIG_DEBUG_WX=y
 CONFIG_DEBUG_STACK_USAGE=y
diff --git a/gki/x86_64/abi.stg b/gki/x86_64/abi.stg
index 68c747a6d02c..c1017f85ff5e 100644
--- a/gki/x86_64/abi.stg
+++ b/gki/x86_64/abi.stg
@@ -5188,6 +5188,11 @@ pointer_reference {
   kind: POINTER
   pointee_type_id: 0x1311f1a3
 }
+pointer_reference {
+  id: 0x0e546897
+  kind: POINTER
+  pointee_type_id: 0x131144c3
+}
 pointer_reference {
   id: 0x0e562f18
   kind: POINTER
@@ -111863,6 +111868,12 @@ member {
   type_id: 0x9565759f
   offset: 2688
 }
+member {
+  id: 0xb28a5799
+  name: "margining"
+  type_id: 0x0e546897
+  offset: 7296
+}
 member {
   id: 0xc89058d7
   name: "margins"
@@ -240008,6 +240019,11 @@ struct_union {
   kind: STRUCT
   name: "tb_dma_port"
 }
+struct_union {
+  id: 0x131144c3
+  kind: STRUCT
+  name: "tb_margining"
+}
 struct_union {
   id: 0x0d8ca4ba
   kind: STRUCT
@@ -243252,11 +243268,12 @@ struct_union {
   kind: STRUCT
   name: "usb4_port"
   definition {
-    bytesize: 912
+    bytesize: 920
     member_id: 0xce1ac149
     member_id: 0x48e0c6e1
     member_id: 0x3b2a31d2
     member_id: 0x8833a5f6
+    member_id: 0xb28a5799
   }
 }
 struct_union {
@@ -263262,6 +263279,10 @@ enumeration {
       name: "CODEC_RT1308"
       value: 20
     }
+    enumerator {
+      name: "CODEC_TAS2563"
+      value: 21
+    }
   }
 }
 enumeration {
@@ -276931,6 +276952,14 @@ function {
   parameter_id: 0x33756485
   parameter_id: 0x7f7172c7
 }
+function {
+  id: 0x5c4b7130
+  return_type_id: 0x295c7202
+  parameter_id: 0x3f0185ef
+  parameter_id: 0x3f0185ef
+  parameter_id: 0xf435685e
+  parameter_id: 0x295c7202
+}
 function {
   id: 0x5c4f178f
   return_type_id: 0x18bd6530
@@ -304132,6 +304161,24 @@ elf_symbol {
   type_id: 0x77661da0
   full_name: "crc32_le"
 }
+elf_symbol {
+  id: 0xec1b66f9
+  name: "crc8"
+  is_defined: true
+  symbol_type: FUNCTION
+  crc: 0x80ecca9f
+  type_id: 0x5c4b7130
+  full_name: "crc8"
+}
+elf_symbol {
+  id: 0xac913841
+  name: "crc8_populate_msb"
+  is_defined: true
+  symbol_type: FUNCTION
+  crc: 0xfa29f41b
+  type_id: 0x100ce912
+  full_name: "crc8_populate_msb"
+}
 elf_symbol {
   id: 0xab0e3dee
   name: "crc_ccitt_table"
@@ -325115,7 +325162,7 @@ elf_symbol {
   name: "snd_soc_acpi_intel_detect_amp_type"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x486edcf5
+  crc: 0x311a88e7
   namespace: "SND_SOC_ACPI_INTEL_MATCH"
   type_id: 0xf7516c84
   full_name: "snd_soc_acpi_intel_detect_amp_type"
@@ -325125,7 +325172,7 @@ elf_symbol {
   name: "snd_soc_acpi_intel_detect_codec_type"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x486edcf5
+  crc: 0x311a88e7
   namespace: "SND_SOC_ACPI_INTEL_MATCH"
   type_id: 0xf7516c84
   full_name: "snd_soc_acpi_intel_detect_codec_type"
@@ -325144,7 +325191,7 @@ elf_symbol {
   name: "snd_soc_acpi_intel_get_amp_tplg_suffix"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x70c297bb
+  crc: 0xe914eb3d
   namespace: "SND_SOC_ACPI_INTEL_MATCH"
   type_id: 0xafef0b76
   full_name: "snd_soc_acpi_intel_get_amp_tplg_suffix"
@@ -325154,7 +325201,7 @@ elf_symbol {
   name: "snd_soc_acpi_intel_get_codec_tplg_suffix"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x70c297bb
+  crc: 0xe914eb3d
   namespace: "SND_SOC_ACPI_INTEL_MATCH"
   type_id: 0xafef0b76
   full_name: "snd_soc_acpi_intel_get_codec_tplg_suffix"
@@ -325254,8 +325301,8 @@ elf_symbol {
   name: "snd_soc_acpi_intel_ptl_machines"
   is_defined: true
   symbol_type: OBJECT
-  crc: 0x0acc418c
-  type_id: 0x7caa7325
+  crc: 0x52d6b76e
+  type_id: 0x02c6c6f2
   full_name: "snd_soc_acpi_intel_ptl_machines"
 }
 elf_symbol {
@@ -327541,7 +327588,7 @@ elf_symbol {
   name: "tb_xdomain_alloc_in_hopid"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0xb1c89409
+  crc: 0x554731d1
   type_id: 0x9e261244
   full_name: "tb_xdomain_alloc_in_hopid"
 }
@@ -327550,7 +327597,7 @@ elf_symbol {
   name: "tb_xdomain_alloc_out_hopid"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0xb1c89409
+  crc: 0x554731d1
   type_id: 0x9e261244
   full_name: "tb_xdomain_alloc_out_hopid"
 }
@@ -327559,7 +327606,7 @@ elf_symbol {
   name: "tb_xdomain_disable_paths"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x4aefa20a
+  crc: 0x4fe51fed
   type_id: 0x9e3e5f3b
   full_name: "tb_xdomain_disable_paths"
 }
@@ -327568,7 +327615,7 @@ elf_symbol {
   name: "tb_xdomain_enable_paths"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x4aefa20a
+  crc: 0x4fe51fed
   type_id: 0x9e3e5f3b
   full_name: "tb_xdomain_enable_paths"
 }
@@ -327577,7 +327624,7 @@ elf_symbol {
   name: "tb_xdomain_lane_bonding_disable"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x5b1ecd07
+  crc: 0x1dd2d16d
   type_id: 0x12a223b4
   full_name: "tb_xdomain_lane_bonding_disable"
 }
@@ -327586,7 +327633,7 @@ elf_symbol {
   name: "tb_xdomain_lane_bonding_enable"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x6ab81d72
+  crc: 0xda8e12ca
   type_id: 0x9fba9108
   full_name: "tb_xdomain_lane_bonding_enable"
 }
@@ -327595,7 +327642,7 @@ elf_symbol {
   name: "tb_xdomain_release_in_hopid"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x9ab57a0b
+  crc: 0x58458343
   type_id: 0x133ea0f8
   full_name: "tb_xdomain_release_in_hopid"
 }
@@ -327604,7 +327651,7 @@ elf_symbol {
   name: "tb_xdomain_release_out_hopid"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x9ab57a0b
+  crc: 0x58458343
   type_id: 0x133ea0f8
   full_name: "tb_xdomain_release_out_hopid"
 }
@@ -327613,7 +327660,7 @@ elf_symbol {
   name: "tb_xdomain_request"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x4235bff3
+  crc: 0x9c76fb82
   type_id: 0x9f636cb4
   full_name: "tb_xdomain_request"
 }
@@ -327622,7 +327669,7 @@ elf_symbol {
   name: "tb_xdomain_response"
   is_defined: true
   symbol_type: FUNCTION
-  crc: 0x379fca05
+  crc: 0x0e65759f
   type_id: 0x9f636942
   full_name: "tb_xdomain_response"
 }
@@ -334080,6 +334127,8 @@ interface {
   symbol_id: 0xa49d0aea
   symbol_id: 0x1c25c16f
   symbol_id: 0xd6cf566d
+  symbol_id: 0xec1b66f9
+  symbol_id: 0xac913841
   symbol_id: 0xab0e3dee
   symbol_id: 0xcd2564d1
   symbol_id: 0x2968f72a
diff --git a/gki/x86_64/symbols/desktop b/gki/x86_64/symbols/desktop
index 5c45271ce105..653adbb6f105 100644
--- a/gki/x86_64/symbols/desktop
+++ b/gki/x86_64/symbols/desktop
@@ -369,6 +369,8 @@
   crc16
   crc32_be
   crc32_le
+  crc8
+  crc8_populate_msb
   crc_ccitt_table
   crypto_aead_decrypt
   crypto_aead_encrypt
diff --git a/include/sound/soc-acpi-intel-ssp-common.h b/include/sound/soc-acpi-intel-ssp-common.h
index b4597c8dac78..8ad25b6b3111 100644
--- a/include/sound/soc-acpi-intel-ssp-common.h
+++ b/include/sound/soc-acpi-intel-ssp-common.h
@@ -37,6 +37,9 @@
 #define RT5682_ACPI_HID		"10EC5682"
 #define RT5682S_ACPI_HID	"RTL5682"
 
+/* Ti */
+#define TAS2563_ACPI_HID	"TXNW2563"
+
 enum snd_soc_acpi_intel_codec {
 	CODEC_NONE,
 
@@ -63,6 +66,7 @@ enum snd_soc_acpi_intel_codec {
 	CODEC_RT1015P,
 	CODEC_RT1019P,
 	CODEC_RT1308,
+	CODEC_TAS2563,
 };
 
 enum snd_soc_acpi_intel_codec
diff --git a/sound/soc/codecs/tas2781-i2c.c b/sound/soc/codecs/tas2781-i2c.c
index 089c7da25ab7..3c7e90c05997 100644
--- a/sound/soc/codecs/tas2781-i2c.c
+++ b/sound/soc/codecs/tas2781-i2c.c
@@ -1739,6 +1739,7 @@ static void tasdevice_i2c_remove(struct i2c_client *client)
 
 #ifdef CONFIG_ACPI
 static const struct acpi_device_id tasdevice_acpi_match[] = {
+	{ "TXNW2563", TAS2563 },
 	{ "TAS2781", TAS2781 },
 	{},
 };
diff --git a/sound/soc/intel/boards/Kconfig b/sound/soc/intel/boards/Kconfig
index 4a9324232936..847aef9af63f 100644
--- a/sound/soc/intel/boards/Kconfig
+++ b/sound/soc/intel/boards/Kconfig
@@ -41,6 +41,9 @@ config SND_SOC_INTEL_SOF_CIRRUS_COMMON
 config SND_SOC_INTEL_SOF_NUVOTON_COMMON
 	tristate
 
+config SND_SOC_INTEL_SOF_TI_COMMON
+	tristate
+
 config SND_SOC_INTEL_SOF_BOARD_HELPERS
 	select SND_SOC_ACPI_INTEL_MATCH
 	tristate
@@ -335,11 +338,13 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
 	select SND_SOC_RT5645
 	select SND_SOC_RT5682_I2C
 	select SND_SOC_RT5682S
+	select SND_SOC_TAS2781_I2C
 	select SND_SOC_DMIC
 	select SND_SOC_INTEL_HDA_DSP_COMMON
 	select SND_SOC_INTEL_SOF_BOARD_HELPERS
 	select SND_SOC_INTEL_SOF_MAXIM_COMMON
 	select SND_SOC_INTEL_SOF_REALTEK_COMMON
+	select SND_SOC_INTEL_SOF_TI_COMMON
 	select SND_SOC_ACPI_INTEL_MATCH
 	help
 	   This adds support for ASoC machine driver for SOF platforms
diff --git a/sound/soc/intel/boards/Makefile b/sound/soc/intel/boards/Makefile
index fcd517d6c279..25a1a9066cbf 100644
--- a/sound/soc/intel/boards/Makefile
+++ b/sound/soc/intel/boards/Makefile
@@ -69,5 +69,8 @@ obj-$(CONFIG_SND_SOC_INTEL_SOF_CIRRUS_COMMON) += snd-soc-intel-sof-cirrus-common
 snd-soc-intel-sof-nuvoton-common-y += sof_nuvoton_common.o
 obj-$(CONFIG_SND_SOC_INTEL_SOF_NUVOTON_COMMON) += snd-soc-intel-sof-nuvoton-common.o
 
+snd-soc-intel-sof-ti-common-y += sof_ti_common.o
+obj-$(CONFIG_SND_SOC_INTEL_SOF_TI_COMMON) += snd-soc-intel-sof-ti-common.o
+
 snd-soc-intel-sof-board-helpers-y += sof_board_helpers.o
 obj-$(CONFIG_SND_SOC_INTEL_SOF_BOARD_HELPERS) += snd-soc-intel-sof-board-helpers.o
diff --git a/sound/soc/intel/boards/sof_rt5682.c b/sound/soc/intel/boards/sof_rt5682.c
index 5ceb376d4924..53bf9bc855fe 100644
--- a/sound/soc/intel/boards/sof_rt5682.c
+++ b/sound/soc/intel/boards/sof_rt5682.c
@@ -27,6 +27,7 @@
 #include "sof_board_helpers.h"
 #include "sof_maxim_common.h"
 #include "sof_realtek_common.h"
+#include "sof_ti_common.h"
 
 /* Driver-specific board quirks: from bit 0 to 7 */
 #define SOF_RT5682_MCLK_EN			BIT(0)
@@ -615,6 +616,9 @@ sof_card_dai_links_create(struct device *dev, struct snd_soc_card *card,
 		ctx->amp_link->init = rt5650_spk_init;
 		ctx->amp_link->ops = &sof_rt5682_ops;
 		break;
+	case CODEC_TAS2563:
+		sof_tas2563_dai_link(dev, ctx->amp_link);
+		break;
 	default:
 		dev_err(dev, "invalid amp type %d\n", ctx->amp_type);
 		return -EINVAL;
@@ -756,6 +760,7 @@ static int sof_audio_probe(struct platform_device *pdev)
 	case CODEC_MAX98360A:
 	case CODEC_RT1019P:
 	case CODEC_RT5650:
+	case CODEC_TAS2563:
 	case CODEC_NONE:
 		/* no codec conf required */
 		break;
@@ -892,6 +897,14 @@ static const struct platform_device_id board_ids[] = {
 					SOF_SSP_PORT_BT_OFFLOAD(2) |
 					SOF_BT_OFFLOAD_PRESENT),
 	},
+	{
+		.name = "ptl_tas2563_rt5682",
+		.driver_data = (kernel_ulong_t)(SOF_RT5682_MCLK_EN |
+					SOF_SSP_PORT_CODEC(0) |
+					SOF_SSP_PORT_AMP(1) |
+					SOF_SSP_PORT_BT_OFFLOAD(2) |
+					SOF_BT_OFFLOAD_PRESENT),
+	},
 	{ }
 };
 MODULE_DEVICE_TABLE(platform, board_ids);
@@ -916,3 +929,4 @@ MODULE_LICENSE("GPL v2");
 MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_BOARD_HELPERS);
 MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_MAXIM_COMMON);
 MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_REALTEK_COMMON);
+MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_TI_COMMON);
diff --git a/sound/soc/intel/common/soc-acpi-intel-ptl-match.c b/sound/soc/intel/common/soc-acpi-intel-ptl-match.c
index eae75f3f0fa4..21c2e04606d9 100644
--- a/sound/soc/intel/common/soc-acpi-intel-ptl-match.c
+++ b/sound/soc/intel/common/soc-acpi-intel-ptl-match.c
@@ -21,7 +21,20 @@ static const struct snd_soc_acpi_codecs ptl_rt5682_rt5682s_hp = {
 	.codecs = {RT5682_ACPI_HID, RT5682S_ACPI_HID},
 };
 
+/* The TAS2563/TAS2781 driver uses same driver for management */
+static const struct snd_soc_acpi_codecs ptl_tas2563_amps = {
+	.num_codecs = 1,
+	.codecs = {TAS2563_ACPI_HID}
+};
+
 struct snd_soc_acpi_mach snd_soc_acpi_intel_ptl_machines[] = {
+	{
+		.comp_ids = &ptl_rt5682_rt5682s_hp,
+		.drv_name = "ptl_tas2563_rt5682",
+		.machine_quirk = snd_soc_acpi_codec_list,
+		.quirk_data = &ptl_tas2563_amps,
+		.sof_tplg_filename = "sof-ptl-tas2563-rt5682.tplg",
+	},
 	{
 		.comp_ids = &ptl_rt5682_rt5682s_hp,
 		.drv_name = "ptl_rt5682_def",
